memory_set/src/set.rs

```diff
@@ -137,17 +145,17 @@ impl<F: Copy, P, B: MappingBackend<F, P>> MemorySet<F, P, B> {
         });
 
         // Shrink right if the area intersects with the left boundary.
-        if let Some((before_start, before)) = self.areas.range_mut(..start).last() {
+        if let Some((&before_start, before)) = self.areas.range_mut(..start).last() {
             let before_end = before.end();
             if before_end > start {
                 if before_end <= end {
                     // the unmapped area is at the end of `before`.
-                    before.shrink_right(start.as_usize() - before_start.as_usize(), page_table)?;
+                    before.shrink_right(start.offset_from(before_start) as usize, page_table)?;
                 } else {
                     // the unmapped area is in the middle `before`, need to split.
                     let right_part = before.split(end).unwrap();
-                    before.shrink_right(start.as_usize() - before_start.as_usize(), page_table)?;
-                    assert_eq!(right_part.start(), end);
+                    before.shrink_right(start.offset_from(before_start) as usize, page_table)?;
+                    assert_eq!(right_part.start().into(), Into::<usize>::into(end));
```

Comment:
Why type `B::Addr` can not be compared directly ???

Reply:
because `B::Addr` is not `Debug`